astropy/wcs/utils.py

```diff
@@ -1137,3 +1141,45 @@ def fit_wcs_from_points(xy, world_coords, proj_point='center',
                       np.zeros((degree+1, degree+1)), wcs.wcs.crpix)
 
     return wcs
+
+
+def obsgeo_to_frame(obsgeo, obstime):
+    """
+    Convert a WCS obsgeo property into an ITRS coordinate frame.
+
+    Parameters
+    ----------
+    obsgeo : np.ndarray
+        A shape ``(6, )`` array representing ``OBSGEO-[XYZ], OBSGEO-[BLH]`` as
+        returned by `astropy.wcs.WCS.wcs.obsgeo`.
+
+    obstime : time-like
+        The time assiociated with the coordinate, will be passed to
+        `astropy.coordinates.ITRS` as the obstime keyword.
+
+    Returns
+    -------
+    BaseCoordinateFrame
+        An `~astropy.coordinates.ITRS` coordinate frame representing the coordinates.
+    """
+    invalid_message = (f"Can not parse the obsgeo location ({obsgeo})"
+                       "obsgeo must only be sepecifed in Cartesian or spherical coordinates.")
+
+    if obsgeo is None or np.all(obsgeo == 0) or np.all(~np.isfinite(obsgeo)):
+        raise ValueError(invalid_message)
+
+    data = None
+
+    # If the spherical coords are zero then use the cartesian ones
+    if np.all(obsgeo[3:] == 0):
+        data = CartesianRepresentation(*obsgeo[:3] * u.m)
+
+    # If the cartesian coords are all zero then use the spherical ones
+    elif np.all(obsgeo[:3] == 0):
```

Comment:
If the above suggestion to gather together all "invalid" situations into one `if` is accepted then:
```suggestion
    else np.all(obsgeo[:3] == 0):
```